### ISSUE TITLE
C++ Implementation of data.table::fread

### ISSUE THREAD

**Issue by cca32** [Rdatatable/data.table#5343]
How would I go about using `data.table` directly in C++?

I am huge fan of `data.table` and use it all the time in R. Honestly, it's the main reason why I continue to use R over Python. 

I've been developing some programs in C++ lately and would love to be able to quickly read in large csv files. I have tried the following two solutions:
https://github.com/p-ranav/csv2
https://github.com/ben-strasser/fast-cpp-csv-parser

But I cannot seem to get anything to read in data nearly as fast as data.table can. 

From the following post it seems that data.table memory maps the csv files then uses C to quickly parse the information:
https://stackoverflow.com/questions/24424361/reason-behind-speed-of-fread-in-data-table-package-in-r

as mentioned `verbose=TRUE` helps understand whats going on a little bit more
```
> fread("test.csv",verbose=TRUE)
Input contains no \n. Taking this to be a filename to open
File opened, filesize is 0.486 GB
File is opened and mapped ok
Detected eol as \n only (no \r afterwards), the UNIX and Mac standard.
Using line 30 to detect sep (the last non blank line in the first 'autostart') ... sep=','
Found 6 columns
First row with 6 fields occurs on line 1 (either column names or first row of data)
All the fields on line 1 are character fields. Treating as the column names.
Count of eol after first data row: 10000001
Subtracted 1 for last eol and any trailing empty lines, leaving 10000000 data rows
Type codes (   first 5 rows): 113431
Type codes (+ middle 5 rows): 113431
Type codes (+   last 5 rows): 113431
Type codes: 113431 (after applying colClasses and integer64)
Type codes: 113431 (after applying drop or select (if supplied)
Allocating 6 column slots (6 - 0 dropped)
Read 10000000 rows and 6 (of 6) columns from 0.486 GB file in 00:00:44
  13.420s ( 31%) Memory map (rerun may be quicker)
   0.000s (  0%) sep and header detection
   3.210s (  7%) Count rows (wc -l)
   0.000s (  0%) Column type detection (first, middle and last 5 rows)
   1.310s (  3%) Allocation of 10000000x6 result (xMB) in RAM
  25.580s ( 59%) Reading data
   0.000s (  0%) Allocation for type bumps (if any), including gc time if triggered
   0.000s (  0%) Coercing data already read in type bumps (if any)
   0.040s (  0%) Changing na.strings to NA
  43.560s        Total
  ``` 
  
  is it possible to compile the relevant C files into my C++ project? files such as:
  `dt_stdio.h, fread.c, fread.h, myomp.h`
  is there some easier interface outside of R directly to access `fread` function? 
  
  will `data.table` be available in other languages outside of R? I have seen Python have a package based off this `datatable`

**Comment by cca32** [comment]
I have an `include` folder including all the relevant files from my R install `R/R-3.6.1/include` as well as the following files from `data.table`: `dt_stdio.h fread.h freadLookups.h freadR.c freadR.h po.h fread.c`

I have a `main.cpp`:
```
extern "C" {
    #include <fread.h>
}
#include <iostream>
int main(int argc, char* argv[]) {
    return 0;
}
```
I compile the data.table C code:
`gcc -Iinclude -c -o fread.o fread.c`
I compile the C++ code:
`g++-Iinclude -c -o main.o main.cpp`
I link the two:
`g++ -Iinclude -o main.exe main.o fread.o`

Upon linking however, I get the following errors which I cannot figure out yet:
```
x86_64-w64-mingw32/bin/ld.exe: fread.o:fread.c:(.text+0x1d0): undefined reference to `libintl_dgettext'
x86_64-w64-mingw32/bin/ld.exe: fread.o:fread.c:(.text+0x1da): undefined reference to `Rprintf'
x86_64-w64-mingw32/bin/ld.exe: fread.o:fread.c:(.text+0x42e): undefined reference to `libintl_dgettext'
x86_64-w64-mingw32/bin/ld.exe: fread.o:fread.c:(.text+0x447): undefined reference to `__halt'
```

I will keep trying

**Comment by MichaelChirico** [comment]
dgettext is from gettext, you'll need <libintl.h> (see po.h -- you'll also need -DENABLE_NLS)

I'm not sure where you're getting Rprintf from, and I'm not sure where __halt is from. data.table might have some layering issue going on here that's worth fixing. e.g. we probably need to #include po.h in fread.h (I'm not seeing it at a glance)

**Comment by cca32** [comment]
`fread.h` has a `#include "freadR.h"` which then has `#include "po.h"` and i see that `<libintl.h>` gets called there if `ENABLE_NLS` is defined. and it seems it gets defined in `Rconfig.h` from `R` include files

**Comment by ben-schwen** [comment]
You might also want to check out the [fastConcat](https://github.com/msummersgill/fastConcat) package which is essentially extracting parts of fread and fwrite into a separate package for fast string concatenation.

There it should at least be visible which parts are needed for fread / fwrite only.

**Comment by MichaelChirico** [comment]
> `fread.h` has a `#include "freadR.h"` which then has `#include "po.h"` and i see that `<libintl.h>` gets called there if `ENABLE_NLS` is defined. and it seems it gets defined in `Rconfig.h` from `R` include files

I see... probably better for us to better isolate `fread.h` from R code. `-DDT_PY` might help there, but we could probably better improve it.

**Comment by cca32** [comment]
defining `DTPY` changes the `freadR.h` call to `py_fread.h` call. but i cannot seem to find any `py_fread.h` file
https://github.com/Rdatatable/data.table/search?q=py_fread
```
#ifndef dt_FREAD_H
#define dt_FREAD_H
#include "dt_stdio.h"  // PRId64
#include <stdint.h>    // uint32_t
#include <stdlib.h>    // size_t
#include <stdbool.h>   // bool
#include "myomp.h"
#ifdef DTPY
  #include "py_fread.h"
  #define ENC2NATIVE(s) (s)
#else
  #include "freadR.h"
  extern cetype_t ienc;
  // R's message functions only take C's char pointer not SEXP, where encoding info can't be stored
  // so must convert the error message char to native encoding first in order to correctly display in R
  #define ENC2NATIVE(s) translateChar(mkCharCE(s, ienc))
#endif
```

**Comment by codymarquart** [comment]
@cca32 did you ever make more progress on this?  I'm trying something similar (including data.table in a c++ project) and encountered the same errors.  And I'm also not having any luck locating a `py_fread.h`, so the `DT_PY` flag still doesn't help

**Comment by cca32** [comment]
@codymarquart nope, never figured it out. Wound up just converting CSVs to serialized C++ objects for faster read in C++. If you find a solution or get further, please let me know!

**Comment by dereckmezquita** [comment]
I'm also interested in leveraging `data.table` in a separate project; NodeJS. If anyone could guide us please help.
